### ISSUE TITLE
resnet50 imagenet weights are differents in tf.keras vs keras

### ISSUE THREAD

**Issue by babaish** [tensorflow/tensorflow#33459]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): win10, pycharm (same issue from colab)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pycharm
- TensorFlow version (use command below): 2.0 and keras 2.3.0 (same issue with tf 1.15 and keras 2.2.5 in google colab)
- Python version: 3.7

**Describe the current behavior**
The weights of the kernel in layers 13 & 14 are different in tf.keras and keras: the weights of the layer 13 in tf.keras are the ones of the layers 14 in keras and the weights of the layer 14 in tf.keras are the ones of the layer 13 in keras:
with tf.keras:
layer 13: conv2_block1_0_conv
[[[[ 0.00460704  0.0613995   0.04595907 ... -0.12616335 -0.00781816
     0.03271283]
   ...
   [-0.00736084  0.00832207 -0.00591875 ... -0.16227128  0.00581011
     0.01718325]]]]
layer 14: conv2_block1_3_conv
[[[[ 0.00412396 -0.01779881 -0.01002417 ... -0.0397268  -0.01897338
    -0.00012411]
   ...
   [ 0.01601992  0.00197976 -0.01605847 ...  0.06464136  0.0353195
     0.02405972]]]]

with keras:
layer 13: res2a_branch2c
[[[[ 0.00412396 -0.01779881 -0.01002417 ... -0.0397268  -0.01897338
    -0.00012411]
   ...
   [ 0.01601992  0.00197976 -0.01605847 ...  0.06464136  0.0353195
     0.02405972]]]]
layer 14: res2a_branch1
[[[[ 0.00460704  0.0613995   0.04595907 ... -0.12616335 -0.00781816
     0.03271283]
   ...
   [-0.00736084  0.00832207 -0.00591875 ... -0.16227128  0.00581011
     0.01718325]]]]

this is with include_top=True but include_top=False gives the same thing

**Describe the expected behavior**
I would expect the same weights, not sure which weights are the right ones

**Code to reproduce the issue**
from tensorflow.python.keras.applications import ResNet50
#from keras.applications import ResNet50
image_size = 224
model = ResNet50(input_shape=(image_size, image_size, 3), include_top=True, weights='imagenet')
print(model.layers[13].name)
weights = model.layers[13].get_weights()[0]
print(weights)
print(model.layers[14].name)
weights = model.layers[14].get_weights()[0]
print(weights)

**Other info / logs**
The difference is coming from the different path used to load the h5 file:
in tf.keras:
'https://github.com/keras-team/keras-applications/releases/download/resnet/resnet50_weights_tf_dim_ordering_tf_kernels.h5'
in keras: 
'https://github.com/fchollet/deep-learning-models/releases/download/v0.2/resnet50_weights_tf_dim_ordering_tf_kernels.h5'

**Comment by gadagashwini** [comment]
I could reproduce the issue with Tf 2.0.0 keras 2.3.0 and TF 1.15.0 keras 2.2.5. Please take a look at the gist of tf.keras [here](https://colab.sandbox.google.com/gist/gadagashwini/903f26e451e5b6d40ac5f99991716a25/untitled206.ipynb) and keras with tensorflow backend [here](https://colab.sandbox.google.com/gist/gadagashwini/99de5876f3f6ff5937bb29c6ab1185e7/untitled207.ipynb). Thanks!

**Comment by indranaut** [comment]
@gadagashwini This does not answer anything. The issue still remains. Or is it that they are downloading two versions of resnet (v1 and v2) ?

**Comment by qlzh727** [comment]
Adding Francois for this. When did we update the weights last time?

**Comment by netw0rkf10w** [comment]
Any updates on this please?

**Comment by netw0rkf10w** [comment]
It seems that several blocks are affected. For example, in the `conv5_` block the weights of the following layers are switched:
- `conv5_block1_3_bn` and `conv5_block1_0_bn`
- `conv5_block1_3_conv` and `conv5_block1_0_conv`.

This affects quite significantly the accuracy of transfer learning (on at least semantic segmentation in my case).

**Comment by CNOCycle** [comment]
I guess #35336 is a similar issue.

In my testing code which loading weights from tf.keras, the accuracy for pre-trained weights are under 70%.

**Comment by netw0rkf10w** [comment]
@CNOCycle Nice. I guess we now have enough evidence that the pre-trained weights are broken.

**Comment by qlzh727** [comment]
I think tf.keras.application is exporting the latest implementation of resnet50, while keras.application is exporting the old version of implementation. They are structured differently (eg the layer may appear in different order), but functionality wise, they are the same.

I think the root cause is for keras.application not exporting the latest version is in https://github.com/keras-team/keras-applications/blob/master/keras_applications/__init__.py. Basically both resnet50.py and resnet.py will export Resnet50, and the former one is exported by contains the legacy implementation.

**Comment by qlzh727** [comment]
Checked with @fchollet offline for this issue. I think the keras-team/keras-application was exporting the old model. The PR should fix the issue, but since the keras-application is not going to make any new release, I would suggest you to use the version in tf.keras.application, which should be latest and correct.

Btw, the keras-team/keras-application is still correct, it is just structured differently compare to the latest implementation. You can still use it for transfer learning.

**Comment by netw0rkf10w** [comment]
Not sure I fully understood the root cause, but I'm happy there's some progress on this.

The difference between tf.keras and keras does not seem to be an issue as they used two different implementations. But is it sure that the weights of tf.keras' ResNet50 are good? As @CNOCycle pointed out in #35336, the best accuracy he/she could get on ImageNet validation set is 45.7% for ResNet50, which is far from the [reported](https://keras.io/applications/) 74.9%.

@CNOCycle Would it be possible for you to try applying the above PR and performing your experiment again?

**Comment by CNOCycle** [comment]
I do the test again in docker image: `tensorflow/tensorflow:2.1.0-gpu-py3`

and comment out the line as @qlzh727 suggestion

`/usr/local/lib/python3.6/dist-packages/keras_applications/__init__.py" 65L`
```diff
-from . import resnet50
+#from . import resnet50
```

The following is the output of testing code which is shown in #35336
```
Downloading data from https://github.com/keras-team/keras-applications/releases/download/resnet/resnet50_weights_tf_dim_ordering_tf_kernels.h5
102973440/102967424 [==============================] - 204s 2us/step
2020-03-02 12:37:29.774845: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2020-03-02 12:37:30.223375: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
500/500 [==============================] - 143s 287ms/step - loss: 2.6850 - accuracy: 0.4601
[Testing][pixel vales are from (0,255)][model:ResNet50] - loss: 2.685 - accuracy: 0.460
Downloading data from https://github.com/keras-team/keras-applications/releases/download/densenet/densenet121_weights_tf_dim_ordering_tf_kernels.h5
33193984/33188688 [==============================] - 107s 3us/step
500/500 [==============================] - 80s 159ms/step - loss: 39.1680 - accuracy: 0.0056
[Testing][pixel vales are from (0,255)][model:DenseNet121] - loss: 39.168 - accuracy: 0.006
Downloading data from https://github.com/JonathanCMitchell/mobilenet_v2_keras/releases/download/v1.1/mobilenet_v2_weights_tf_dim_ordering_tf_kernels_1.0_224.h5
14540800/14536120 [==============================] - 39s 3us/step
500/500 [==============================] - 41s 81ms/step - loss: 9.9651 - accuracy: 0.0026
[Testing][pixel vales are from (0,255)][model:MobileNetV2] - loss: 9.965 - accuracy: 0.003
500/500 [==============================] - 76s 151ms/step - loss: 8.5355 - accuracy: 0.0010
[Testing][pixel vales are from (0,1)][model:ResNet50] - loss: 8.535 - accuracy: 0.001
500/500 [==============================] - 71s 142ms/step - loss: 1.7376 - accuracy: 0.6029
[Testing][pixel vales are from (0,1)][model:DenseNet121] - loss: 1.738 - accuracy: 0.603
500/500 [==============================] - 45s 91ms/step - loss: 2.2516 - accuracy: 0.5285
[Testing][pixel vales are from (0,1)][model:MobileNetV2] - loss: 2.252 - accuracy: 0.528
500/500 [==============================] - 76s 152ms/step - loss: 8.4942 - accuracy: 0.0010
[Testing][pixel vales are normalized from (-1,1)][model:ResNet50] - loss: 8.494 - accuracy: 0.001
500/500 [==============================] - 71s 142ms/step - loss: 1.9044 - accuracy: 0.5703
[Testing][pixel vales are normalized from (-1,1)][model:DenseNet121] - loss: 1.904 - accuracy: 0.570
500/500 [==============================] - 46s 93ms/step - loss: 2.4290 - accuracy: 0.4992
[Testing][pixel vales are normalized from (-1,1)][model:MobileNetV2] - loss: 2.429 - accuracy: 0.499
```

However, I still get similar results. I hope that someone would check correctness of my code.

**Comment by qlzh727** [comment]
@CNOCycle, note that the downloading url is from https://github.com/keras-team/keras-applications/releases, which should be same as the one in tf.keras.application. This somehow indicating that your transfer learning model might not be constructed correctly.

The easiest way to verify whether the original resnet model is correct is to run some eval with the eval data, which I am quite sure we did before we release the model.

**Comment by netw0rkf10w** [comment]
> @CNOCycle, note that the downloading url is from https://github.com/keras-team/keras-applications/releases, which should be same as the one in tf.keras.application. This somehow indicating that your transfer learning model might not be constructed correctly.

Right. I have just tried again with `tf-nightly`, and indeed, the URLs have changed:

> Downloading data from https://storage.googleapis.com/tensorflow/keras-applications/resnet/resnet50_weights_tf_dim_ordering_tf_kernels_notop.h5

One can check this in my [Colab notebook](https://colab.research.google.com/drive/13zpJSsJ0GmXluiVNAF4DLSqW3Ao84G5Z).

@CNOCycle Could you please try again with `tf-nightly`? Thanks a lot.

**Comment by CNOCycle** [comment]
I checked the weight from `storage.googleapis.com` and `keras-applications/releases`. Both files' hash are same.

```
2cb95161c43110f7111970584f804107  /root/.keras/models/resnet50_weights_tf_dim_ordering_tf_kernels.h5
```

What I do is not transfer learning. I'm trying to reproduce to accuracy which publish in many papers. 

From my experience, my implement of training ImageNet from scratch by TF and pytorch. For pytorch version, its top-1 accuracy is about 74%. For TF version, its top-1 accuracy is about 68%.  For TF version with pre-trained weights, its top-1 accuracy is under 50%

**Comment by netw0rkf10w** [comment]
Hmm, that's bad news... 

> The easiest way to verify whether the original resnet model is correct is to run some eval with the eval data, which I am quite sure we did before we release the model.

This is exactly what @CNOCycle did. @qlzh727 Would you mind sharing your eval code so that we can check the weights again on our side?

**Comment by netw0rkf10w** [comment]
Hmm... I think this issue is rather critical... Imagine getting 2-3% lower accuracy than PyTorch just because of bad conversion of pretrained models? That'd be a pity...

**Comment by nick-nikzad** [comment]
Hello everyone,
I have just tried to use pre-trained ResNet50 with tf.keras.applications.ResNet50(**args). The thing is happening for me is that after downloading weights, nothing else happens!! even after a couple of hours it does not go any further!! Any idea? what may be the casue of this? Any solution?

Many thanks for your help in advance.

I use as follow:

ENCODER_BASE =  tf.keras.applications.ResNet50(include_top=False,input_shape=(None,None,3), weights='imagenet')

**Comment by netw0rkf10w** [comment]
@nick-nikzad At least show us your code to see what you would expect to see after loading the model.

**Comment by nick-nikzad** [comment]
@netw0rkf10w  Thanks for the reply. I just expect using the output of some blocks as below: 
    
    ENCODER_BASE=tf.keras.applications.ResNet50(include_top=False,input_shape=(None,None,3), weights='imagenet')
    for layer in ENCODER_BASE.layers:
        layer.traiable = True
        
    f0 = ENCODER_BASE.get_layer("conv1_relu").output  #43358  

    f1 = ENCODER_BASE.get_layer("conv2_block2_out").output  
    f2 = ENCODER_BASE.get_layer("conv3_block4_out").output 
    f3 = ENCODER_BASE.get_layer("conv4_block6_out").output  
    f4 = ENCODER_BASE.get_layer("conv5_block3_out").output

**Comment by netw0rkf10w** [comment]
@nick-nikzad Can you reproduce that using Colab?

**Comment by gabriel-vanzandycke** [comment]
https://stackoverflow.com/questions/67365237/imagenet-pretrained-resnet50-backbones-are-different-between-pytorch-and-tensorf

And given my ongoing experiments, transfer learning doesn't work as well in TensorFlow :o
